hedera-node/docs/design/services/smart-contract-service/IHFC-associate-dissociate.md

```diff
@@ -0,0 +1,78 @@
+# IHRC Facade Associate and Dissociate Functionality
+
+## Purpose
+
+[HIP-218](https://hips.hedera.com/hip/hip-218) proposed a way to enable treating HTS tokens in an equivalent manner
+to ERC tokens in the EVM world by creating a proxy redirect facade contract.  We need to extend 
+this functionality to include `associate` and `dissociate` functions which are part of functionality available to HTS tokens
+but not ERC tokens.
+
+## Goals
+
+- Extend the proxy redirect token facade contract to enable calling `associate` and `dissociate` functions on tokens.
+
+## Non Goals
+
+- Refactor the current implementation to better support future additions to the token facade contract functionality.
+
+## Architecture
+
+The method in which the redirection to proxy contract is implemented is that during the execution of an EVM request, we examine the contract address to see
+if it is an HTS token address and if so we redirect the request to the proxy contract.  This mechanism will be extended to include
+calls to `associate` and `dissociate` functions in the HTSPrecompileContact class method which handles ABI_ID_REDIRECT_FOR_TOKEN function selector.
+
+The following table describes the function selector for the new `associate` and `dissociate` functions and the associated function signatures.
+
+| Function Selector | Function Signature |
+|-------------------|--------------------|
+| 0x0a754de6        | associate()        |
+| 0x5c9217e0        | dissociate()       |
+
+No arguments are necessary because 
+- The token address was already determined by looking up the contract address to determine if it is an HTS token address.
+- The address to associate/dissociate will be the caller (i.e. msg.sender)
+
+Once the above functionality has been implemented in services, the end user will be able to call the `associate` and `dissociate` functions as follows:
+
+```
+IHRC(tokenAddress).associate()
+IHRC(tokenAddress).dissociate()
+```
+
+The solidity interface for IHRC will be the following
+
+```
+interface IHRC {
+    function associate() external returns (bool);
```

Comment:
Q: Should these be response codes rather?
```suggestion
    function associate() external returns (int responseCode);
```

Reply:
That would be more consistent with the other HTS calls but the ERC calls seems to just return a boolean.  responseCode would be easier to deal with because the default does that.

Reply:
+1 for the `int responseCode` return. 

IMO it is more useful for users to know exactly what failed ( the `associate` can fail due to numerous different reasons) instead of returning `false`. Also, as you've pointed out,  `IHederaTokenService#associateToken[s]` already returns response code

Reply:
OK.  I'll update so that we return the response code